### PR TITLE
[EasyAPiToken] Update firebase/php-jwt

### PR DESCRIPTION
**Pull request by alexndlm**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes

Due to https://github.com/advisories/GHSA-8xf4-w7qw-pjjw

See https://github.com/firebase/php-jwt/releases/tag/v6.0.0

### CODE REVIEW

**Review by natepage — CHANGES_REQUESTED** [review]
There is a few BC breaks in this PR, can we please try not to introduce any and support both php-jwt v5 and v6, otherwise we will have to wait for the next major release to get that in

File: packages/EasyApiToken/src/External/Auth0/TokenGenerator.php [hunk 1]
```diff
@@ -76,6 +76,6 @@ public function generate(
             true
         ) : $this->secret;
 
-        return JWT::encode($payload, (string)$secret);
+        return JWT::encode($payload, (string)$secret, 'HS256');
```
Comment (itorgov) [inline packages/EasyApiToken/src/External/Auth0/TokenGenerator.php:79]: Please use the `EonX\EasyApiToken\Interfaces\AlgorithmsInterface::HS256` constant.

File: packages/EasyApiToken/src/External/AwsCognito/JwkFetcher.php [hunk 2]
```diff
@@ -93,7 +74,7 @@ private function fetchKeysFromAws(UserPoolConfigInterface $userPoolConfig): arra
         $keys = [];
 
         foreach ($response['keys'] as $jwk) {
-            $keys[$jwk['kid']] = $this->convertJwkToPem($jwk);
+            $keys[(string)$jwk['kid']] = new Key($this->convertJwkToPem($jwk), $jwk['alg'] ?? 'HS256');
```
Comment (itorgov) [inline packages/EasyApiToken/src/External/AwsCognito/JwkFetcher.php:77]: The same here, the `\EonX\EasyApiToken\Interfaces\AlgorithmsInterface::HS256` const please.

**Review by itorgov — CHANGES_REQUESTED** [review]

**Comment by itorgov** [comment]
> There is a few BC breaks in this PR, can we please try not to introduce any and support both php-jwt v5 and v6, otherwise we will have to wait for the next major release to get that in



> There is a few BC breaks in this PR, can we please try not to introduce any and support both php-jwt v5 and v6, otherwise we will have to wait for the next major release to get that in

Agree 🙂

File: packages/EasyApiToken/src/External/AwsCognito/JwkFetcher.php (hunk 2 above)
Comment (alexndlm) [reply packages/EasyApiToken/src/External/AwsCognito/JwkFetcher.php]: Refactored.

File: packages/EasyApiToken/src/External/AwsCognitoJwtDriver.php [hunk 3]
```diff
@@ -11,30 +11,33 @@
...
      */
     public function __construct(
-        private UserPoolConfigInterface $userPoolConfig,
-        ?JwkFetcherInterface $jwkFetcher = null,
+        private readonly UserPoolConfigInterface $userPoolConfig,
+        private JwkFetcherInterface $jwkFetcher = new JwkFetcher(),
         ?array $allowedAlgos = null,
-        private ?int $leeway = null
+        private readonly ?int $leeway = null,
+        private readonly string $defaultJwkAlgo = 'HS256'
```
Comment (ERuban) [inline packages/EasyApiToken/src/External/AwsCognitoJwtDriver.php:39]: Shouldn't it be `RS256` by default for AwsCognitoJwtDriver? As I can see, we using `RS256` for Cognito JWT.
And, please, use the constant if possible

File: packages/EasyApiToken/src/External/AwsCognitoJwtDriver.php [hunk 4]
```diff
@@ -48,8 +51,16 @@ public function decode(string $token): object
             JWT::$leeway = $this->leeway;
         }
 
-        /** @var \stdClass $decodedToken */
-        $decodedToken = JWT::decode($token, $this->jwkFetcher->getJwks($this->userPoolConfig), $this->allowedAlgos);
+        $jwks = $this->jwkFetcher->getJwks($this->userPoolConfig);
+
+        if (self::isFirebaseJwtV5() === false) {
+            foreach ($jwks as $keyId => $key) {
+                $jwks[$keyId] = new Key($key, $this->defaultJwkAlgo ?? 'HS256');
```
Comment (ERuban) [inline packages/EasyApiToken/src/External/AwsCognitoJwtDriver.php:58]: `$this->defaultJwkAlgo` is not nullable

File: packages/EasyApiToken/src/External/JwtTrait.php [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+     */
+    private static function isFirebaseJwtV5(): bool
+    {
+        if (isset(self::$isFirebaseJwtV5) === false) {
+            $methodParams = \array_map(
+                static fn ($parameter) => $parameter->name,
+                (new ReflectionMethod(BaseJWT::class, 'decode'))->getParameters()
+            );
+
+            self::$isFirebaseJwtV5 = \in_array('allowed_algs', $methodParams, true);
```
Comment (ERuban) [inline packages/EasyApiToken/src/External/JwtTrait.php:30]: `allowed_algs` -> `allowed_algos`?

**Review by ERuban — CHANGES_REQUESTED** [review]

File: packages/EasyApiToken/src/External/JwtTrait.php (hunk 5 above)
Comment (alexndlm) [reply packages/EasyApiToken/src/External/JwtTrait.php:30]: Here ok - https://github.com/firebase/php-jwt/blob/83b609028194aa042ea33b5af2d41a7427de80e6/src/JWT.php#L83

File: packages/EasyApiToken/src/External/AwsCognitoJwtDriver.php (hunk 3 above)
Comment (alexndlm) [reply packages/EasyApiToken/src/External/AwsCognitoJwtDriver.php]: Yes, you are right. Refactored.

File: packages/EasyApiToken/src/External/AwsCognitoJwtDriver.php (hunk 4 above)
Comment (alexndlm) [reply packages/EasyApiToken/src/External/AwsCognitoJwtDriver.php]: Thanks. Fixed.

File: packages/EasyApiToken/src/External/Auth0/TokenGenerator.php [hunk 6]
```diff
@@ -9,30 +9,18 @@
...
-     *
-     * @var string|null
-     */
-    private $audience;
-
-    /**
-     * @var null|string
-     */
-    private $issuer;
+    private const DEFAULT_ALG = 'HS256';
```
Comment (ERuban) [inline packages/EasyApiToken/src/External/Auth0/TokenGenerator.php:12]: `DEFAULT_ALGO`? :)

**Review by ERuban — CHANGES_REQUESTED** [review]

File: packages/EasyApiToken/src/External/Auth0/TokenGenerator.php (hunk 6 above)
Comment (alexndlm) [reply packages/EasyApiToken/src/External/Auth0/TokenGenerator.php]: Yes. Thanks

File: packages/EasyApiToken/src/External/JwtTrait.php [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+namespace EonX\EasyApiToken\External;
+
+use Firebase\JWT\JWT as BaseJWT;
+use ReflectionMethod;
+
+/**
+ * @deprecated compatibility layer, will be removed in 5.0
+ * @internal
+ */
+trait JwtTrait
```
Comment (ERuban) [inline packages/EasyApiToken/src/External/JwtTrait.php:14]: Let's name it a bit more obviously?
Dunno, maybe smth like `FirebaseJwtVersionControlTrait`, or just `FirebaseJwtVersionTrait`?
wdyt?

File: packages/EasyApiToken/src/External/AwsCognito/JwkFetcher.php [hunk 8]
```diff
@@ -17,33 +17,13 @@
 
 final class JwkFetcher implements JwkFetcherInterface
 {
-    /**
-     * @var int
-     */
-    private const CACHE_EXPIRY = 3600;
```
Comment (itorgov) [inline packages/EasyApiToken/src/External/AwsCognito/JwkFetcher.php:23]: Why did you delete it? Let's just rename to `DEFAULT_CACHE_EXPIRY`.

File: packages/EasyApiToken/src/External/JwtTrait.php [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+    /**
+     * @deprecated compatibility layer, will be removed in 5.0
+     * @internal
+     */
+    private static function isFirebaseJwtV5(): bool
+    {
+        if (isset(self::$isFirebaseJwtV5) === false) {
+            $methodParams = \array_map(
+                static fn ($parameter) => $parameter->name,
+                (new ReflectionMethod(BaseJWT::class, 'decode'))->getParameters()
```
Comment (ERuban) [inline packages/EasyApiToken/src/External/JwtTrait.php:27]: Hm, maybe it would be better to use smth like `Composer\InstalledVersions::getVersion($packageName)`?
I don't like to have reflection usage on each request.
wdyt?

**Review by ERuban — CHANGES_REQUESTED** [review]

**Review by itorgov — CHANGES_REQUESTED** [review]

File: packages/EasyApiToken/src/External/JwtTrait.php (hunk 9 above)
Comment (alexndlm) [reply packages/EasyApiToken/src/External/JwtTrait.php:27]: Then we will need extra packages.
@natepage do you mind?

Снимок экрана 2023-02-20 в 16 21 16

Снимок экрана 2023-02-20 в 16 21 03

Comment (ERuban) [reply packages/EasyApiToken/src/External/JwtTrait.php:27]: We don't need these packages. You can use `version_compare` PHP function to compare version you got from `Composer\InstalledVersions::getVersion($packageName)`.

File: packages/EasyApiToken/src/External/JwtTrait.php (hunk 7 above)
Comment (akl-95) [reply packages/EasyApiToken/src/External/JwtTrait.php:14]: I like `FirebaseJwtVersionTrait`.

File: packages/EasyApiToken/src/External/JwtTrait.php (hunk 9 above)
Comment (alexndlm) [reply packages/EasyApiToken/src/External/JwtTrait.php:27]: At least we need `composer-runtime-api` for `Composer\InstalledVersions::getVersion($packageName)`

File: packages/EasyApiToken/src/External/AwsCognitoJwtDriver.php [hunk 10]
```diff
@@ -48,8 +56,16 @@ public function decode(string $token): object
             JWT::$leeway = $this->leeway;
         }
 
-        /** @var \stdClass $decodedToken */
-        $decodedToken = JWT::decode($token, $this->jwkFetcher->getJwks($this->userPoolConfig), $this->allowedAlgos);
+        $jwks = $this->jwkFetcher->getJwks($this->userPoolConfig);
+
+        if (self::isFirebaseJwtV5() === false) {
```
Comment (akl-95) [inline packages/EasyApiToken/src/External/AwsCognitoJwtDriver.php:61]: This check looks like we are OK with versions <= 4 in this `if`. Let's either do `isFirebaseJwtV5` -> `getFirebaseJwtVersion` or `isFirebaseJwtV5` -> `isFirebaseJwtVersionGt5`. Or anything like that. But I don't like the 2nd option actually, the name looks strange to me.

**Review by akl-95 — CHANGES_REQUESTED** [review]

File: packages/EasyApiToken/src/External/JwtTrait.php (hunk 9 above)
Comment (ERuban) [reply packages/EasyApiToken/src/External/JwtTrait.php:27]: But, looking to its description: 
```
composer-runtime-api 2.0 introduced a new Composer\InstalledVersions class
```
and looking to ours `vendor/composer` folder - we already have it.

Comment (natepage) [reply packages/EasyApiToken/src/External/JwtTrait.php:27]: Usually in such situation I try to identify an easy thing that differentiate 2 versions of the same package, a new class, a missing class, method, etc. So it's quite easy to understand which version of the package you're dealing with by checking if the class exists for example.

Comment (alexndlm) [reply packages/EasyApiToken/src/External/JwtTrait.php]: @ERuban, beyond belief, it works.

File: packages/EasyApiToken/src/External/AwsCognitoJwtDriver.php (hunk 10 above)
Comment (alexndlm) [reply packages/EasyApiToken/src/External/AwsCognitoJwtDriver.php]: Refactored, please check.

File: packages/EasyApiToken/src/External/JwtTrait.php (hunk 7 above)
Comment (alexndlm) [reply packages/EasyApiToken/src/External/JwtTrait.php]: Done

**Review by itorgov — APPROVED** [review]

**Review by ERuban — DISMISSED** [review]

**Review by akl-95 — DISMISSED** [review]

**Review by roman-eonx — APPROVED** [review]

File: packages/EasyApiToken/src/External/AwsCognitoJwtDriver.php [hunk 11]
```diff
@@ -48,8 +56,16 @@ public function decode(string $token): object
             JWT::$leeway = $this->leeway;
         }
 
-        /** @var \stdClass $decodedToken */
-        $decodedToken = JWT::decode($token, $this->jwkFetcher->getJwks($this->userPoolConfig), $this->allowedAlgos);
+        $jwks = $this->jwkFetcher->getJwks($this->userPoolConfig);
+
+        if (self::isFirebaseJwtV6()) {
+            foreach ($jwks as $keyId => $key) {
+                $jwks[$keyId] = new Key($key, $this->defaultJwkAlgo);
```
Comment (natepage) [inline packages/EasyApiToken/src/External/AwsCognitoJwtDriver.php:63]: Are the JWKs already indexed by their keyId from the jwkFetcher? If not then it means that we have an array with each key twice in it, in raw format and as Key object.

File: packages/EasyApiToken/tests/AbstractFirebaseJwtTokenTestCase.php [hunk 12]
```diff
@@ -43,7 +43,7 @@ protected function createFirebaseJwtDriver(
         OpenSSLAsymmetricKey|string|null $publicKey = null,
         OpenSSLAsymmetricKey|string|null $privateKey = null,
         ?array $allowedAlgos = null,
-        ?int $leeway = null
+        ?int $leeway = null,
```
Comment (natepage) [inline packages/EasyApiToken/tests/AbstractFirebaseJwtTokenTestCase.php:46]: We never had trailing comas on function arguments before

**Review by natepage — CHANGES_REQUESTED** [review]

File: packages/EasyApiToken/src/External/AwsCognitoJwtDriver.php (hunk 11 above)
Comment (alexndlm) [reply packages/EasyApiToken/src/External/AwsCognitoJwtDriver.php:63]: `$jwks` is `array<int|string, string>`. How can we have duplication?

File: packages/EasyApiToken/tests/AbstractFirebaseJwtTokenTestCase.php (hunk 12 above)
Comment (alexndlm) [reply packages/EasyApiToken/tests/AbstractFirebaseJwtTokenTestCase.php]: Thanks. Fixed

File: packages/EasyApiToken/src/External/AwsCognitoJwtDriver.php (hunk 11 above)
Comment (natepage) [reply packages/EasyApiToken/src/External/AwsCognitoJwtDriver.php:63]: Because if the original `$jwks` isn't indexed by the keyIds and just by auto-increment integers then we will end up with having the original keys with auto-increments and the key objects indexed by their keyId

Comment (alexndlm) [reply packages/EasyApiToken/src/External/AwsCognitoJwtDriver.php:63]: `\EonX\EasyApiToken\External\AwsCognito\JwkFetcher::fetchKeysFromAws` returns an array of keys, with `kid` as an index.

I can't get what I do wrong.

Comment (natepage) [reply packages/EasyApiToken/src/External/AwsCognitoJwtDriver.php:63]: ok if that's the case then you're not doing anything wrong, I was asking for this confirmation

**Review by natepage — APPROVED** [review]
